doc/warm-reboot/NSF_Manager_HLD.md

```diff
@@ -0,0 +1,292 @@
+
+
+##### Application Shutdown Optimization
+
+
+
+![alt_text](img/shutdown-optimization.png)
+
+Higher layer applications are generally independent and their shutdown is not dependent on quiescence of other switch stack components. For example, upon receiving a freeze notification Teamd can stop exchanging LACP packets, perform state verification (optional) and checkpoint LACP states i.e. transition from Phase 1 to Phase 3 without waiting for NSF Manager to send further notifications. The design allows applications to transition through these phases independently as long as they continue to update their warmboot state in STATE DB. NSF Manager will monitor these states and will handle the applications’ phase transitions. In such a scenario, applications need to set _freeze = true_ and _checkpoint = false_ and thus the freeze notification will result in the application to transition from Phase 1 to Phase 3. 
+
+
+#### Reconciliation Monitoring
+
+
+![alt_text](img/reconciliation.png)
+
+Upon bootup, components will detect warm-boot flag in STATE DB and reconcile to the state before reboot. Components that registered _reconciliation = true_ will update their warm-boot state = RECONCILED in STATE DB after they have completed reconciliation. NSF Manager will monitor the reconciliation status of these registered components during warm bootup. Subsequently, the components will start their normal operation and listen to new requests. The proposed reconciliation monitoring mechanism is similar to the existing mechanism except that components being monitored are based on the NSF details registration as opposed to hardcoded list of components.
+
+
+#### Component Warmboot States
+
+Based on the shutdown and reconciliation sequence mentioned above, a component will transition through the following states during warm reboot:
+
+
+
+*   Frozen (warm shutdown)
+*   Quiescent (warm shutdown)
+*   Checkpointed (optional, warm shutdown)
+*   Initialized (warm bootup)
+*   Reconciled (warm bootup)
+*   Failed (warm shutdown or bootup)
+
+
+
+![alt_text](img/warmboot-states.png)
+
+Upon receiving a freeze notification, a component will transition to _frozen_ state when it has stopped generating new intents for which it is the source. Subsequently, it will transition to _quiescent_ state when it has completed processing all requests. It might transition between _frozen_ and _quiescent_ states when it receives a new request from some other switch component and after it has completed processing all requests i.e. a component might transition between _frozen_ and _quiescent_ states if it processes an intent in freeze mode. If a component fails to perform its freeze routine i.e. fails to stop generating new intents for which it is the source then it will transition to _failed_ state. Upon receiving a checkpoint notification, a component will transition to _checkpointed_ state after it has completed checkpointing or to _failed_ state if checkpointing failed.
```

Comment:
Since this is an initial NSF Manager design, the current proposal is to abort warm reboot if an application reports _failed_ warm-boot state. However, we are exploring `unfreeze` on warm-boot failure as well. For example, if Phase 1 (freeze) fails then we can unfreeze the switch stack and return that warm reboot failed.